### PR TITLE
feat: Check malformed path entries

### LINKED ISSUE #302: the agent should give an error for malformed `path` entries
The agent should check the format of a `path` entry in `appmap.yml`, and fail if it finds a bad one. In particular, it should detect entries that contain path separators (`/`, `\`) and offer a hint that `path` entries are module names, not directory paths.

### PR DESCRIPTION
**Pull request by zermelo-wisen**
Fixes #302

Warning and dropping malformed configuration file package path entries, which cannot be parsed as Python module names: `mod1/mod2`, `mod1\mod2`, `123`, `.`, ` `.

### CODE REVIEW

File: _appmap/configuration.py [hunk 1]
```diff
@@ -116,6 +117,8 @@ def excluded(d):
 
     return packages
 
+class AppMapPackagePathException(Exception):
```
Comment (apotterri) [inline _appmap/configuration.py:120]:
```suggestion
class AppMapInvalidConfigException(Exception):
```

File: _appmap/configuration.py [hunk 2]
```diff
@@ -219,7 +222,8 @@ def _load_config(self):
                         self._config["name"] = self.default_name
                     if "packages" not in self._config:
                         self._config["packages"] = self.default_packages
-
+                    else:
+                        self._check_path_values()
```
Comment (apotterri) [inline _appmap/configuration.py:226]: Having `_check_path_values` raise an exception is good, because it makes it easy to test.

However, showing the user a stack trace here isn't going to be helpful. Instead, this should catch the exception, show the user the message, and call `sys.exit(1)`.

File: _appmap/configuration.py [hunk 3]
```diff
@@ -269,6 +273,24 @@ def _load_functions(self):
...
         self.package_functions.update(modules)
 
+    def _check_path_values(self):
+        for item in self._config["packages"]:
+            path: str = item.get("path")
+            if path and not self._check_path_value(path):
+                has_separator = '/' in path or '\\' in path
+                raise AppMapPackagePathException(
+                    f"Malformed path value '{path}' in configuration file. "
+                    "Path entries should be module names"
```
Comment (apotterri) [inline _appmap/configuration.py:283]:
```suggestion
                    "Path entries must be module names"
```

File: _appmap/configuration.py [hunk 4]
```diff
@@ -269,6 +273,24 @@ def _load_functions(self):
 
         self.package_functions.update(modules)
 
+    def _check_path_values(self):
+        for item in self._config["packages"]:
+            path: str = item.get("path")
+            if path and not self._check_path_value(path):
```
Comment (apotterri) [inline _appmap/configuration.py:279]: This code should also raise an exception if `path` is `None`. This case is currently not handled well, and produces an ugly error message.

**Review by apotterri — CHANGES_REQUESTED** [review]
Thanks for this. Just a couple of minor changes to make, then it'll be good to go.

File: _appmap/configuration.py (hunk 2 above)
Comment (apotterri) [reply _appmap/configuration.py]: What's the value in continuing in this case? The user has made it clear that they'd like to get recordings (by updating their `appmap.yml`), but the change they've made makes that impossible. If it was me, I'd like to learn that immediately, rather than having the message scroll off my screen as all my tests run....

Comment (kgilpin) [reply _appmap/configuration.py]: OK

**Comment by apotterri** [comment]
I gave these changes a try. Because of the way `appmap` gets imported (using a `.pth` file), the UX is terrible:
```
[2024-05-14 06:25:39,785] ERROR _appmap.configuration: Malformed path value '.' in configuration file. Path entries must be module names.
Fatal Python error: init_import_site: Failed to import the site module
Python runtime state: initialized
Traceback (most recent call last):
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 227, in _load_config
    self._check_path_values()
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 294, in _check_path_values
    raise AppMapInvalidConfigException(
_appmap.configuration.AppMapInvalidConfigException: Malformed path value '.' in configuration file. Path entries must be module names.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 617, in <module>
    main()
  File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 610, in main
    execsitecustomize()
  File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 549, in execsitecustomize
    import sitecustomize
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/test/data/fastapi/init/sitecustomize.py", line 1, in <module>
    import appmap
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/appmap/__init__.py", line 9, in <module>
    from _appmap import generation  # noqa: F401
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/__init__.py", line 1, in <module>
    from . import configuration
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 450, in <module>
    initialize()
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 445, in initialize
    Config().initialize()
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 146, in __init__
    self._load_config()
  File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 230, in _load_config
    sys.exit(1)
SystemExit: 1
```

So, we need to change this. When an invalid `path` is detected`, we should just log a warning and ignore it. This needs to happen without raising an exception, and the agent should not exit.

**Review by apotterri — CHANGES_REQUESTED** [review]
https://github.com/getappmap/appmap-python/pull/303#issuecomment-2109843178

**Comment by zermelo-wisen** [comment]
> I gave these changes a try. Because of the way `appmap` gets imported (using a `.pth` file), the UX is terrible:
> 
> ```
> [2024-05-14 06:25:39,785] ERROR _appmap.configuration: Malformed path value '.' in configuration file. Path entries must be module names.
> Fatal Python error: init_import_site: Failed to import the site module
> Python runtime state: initialized
> Traceback (most recent call last):
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 227, in _load_config
>     self._check_path_values()
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 294, in _check_path_values
>     raise AppMapInvalidConfigException(
> _appmap.configuration.AppMapInvalidConfigException: Malformed path value '.' in configuration file. Path entries must be module names.
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 617, in <module>
>     main()
>   File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 610, in main
>     execsitecustomize()
>   File "/Users/ajp/.asdf/installs/python/3.10.13/lib/python3.10/site.py", line 549, in execsitecustomize
>     import sitecustomize
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/test/data/fastapi/init/sitecustomize.py", line 1, in <module>
>     import appmap
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/appmap/__init__.py", line 9, in <module>
>     from _appmap import generation  # noqa: F401
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/__init__.py", line 1, in <module>
>     from . import configuration
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 450, in <module>
>     initialize()
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 445, in initialize
>     Config().initialize()
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 146, in __init__
>     self._load_config()
>   File "/Users/ajp/src/applandinc/worktrees/appmap-python/_appmap/configuration.py", line 230, in _load_config
>     sys.exit(1)
> SystemExit: 1
> ```
> 
> So, we need to change this. When an invalid `path` is detected`, we should just log a warning and ignore it. This needs to happen without raising an exception, and the agent should not exit.

Done

File: _appmap/configuration.py [hunk 5]
```diff
@@ -269,6 +274,40 @@ def _load_functions(self):
...
+                has_separator = isinstance(path, str) and ('/' in path or '\\' in path)
+                logger.warning(
+                    f"Malformed path value '{path}' in configuration file. "
+                    "Path entries must be module names"
+                    f"{' not directory paths' if has_separator else ''}."
+                )
+                invalid_items.append(item)
+                continue
+
+        if len(invalid_items) > 0:
```
Comment (apotterri) [inline _appmap/configuration.py:300]: What's going to happen if this eliminates all the path entries?

Comment (zermelo-wisen) [reply _appmap/configuration.py:300]: I think, no package will be instrumented, unless there are some `dist` entries. We reach this state with at least one warning.

Thinking again, should we check if:
 1. `packages` is a yaml sequence and not a scalar
 2. `packages` is empty
 3. Every item has `path` XOR `dist` key

Comment (apotterri) [reply _appmap/configuration.py:300]: No, let's not expand the scope of this beyond checking that `path`s are modules.

However, please add a test that has only invalid `path`s, and confirm that the resulting `Config` has a `package` with an empty list of `paths`.

Comment (zermelo-wisen) [reply _appmap/configuration.py:334]: Done

**Review by apotterri — APPROVED** [review]
